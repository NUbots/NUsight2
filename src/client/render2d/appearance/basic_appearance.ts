import { observable } from 'mobx'

type Fill = { color: string, alpha: number }
type Stroke = { color: string, alpha: number, width: number }

export type BasicAppearanceOpts = {
  fill?: Partial<Fill>,
  stroke?: Partial<Stroke>
}

export class BasicAppearance {
  @observable fill?: Fill
  @observable stroke?: Stroke

  constructor({ fill, stroke }: { fill?: Fill, stroke?: Stroke }) {
    this.fill = fill
    this.stroke = stroke
  }

  static of({ fill, stroke }: BasicAppearanceOpts = {}): BasicAppearance {
    return new BasicAppearance({
      fill: fill && { color: '#000000', alpha: 1, ...fill },
      stroke: stroke && { color: '#000000', alpha: 1, width: 1, ...stroke },
    })
  }
}
