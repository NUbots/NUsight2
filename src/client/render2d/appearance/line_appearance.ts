import { observable } from 'mobx'


export type LineAppearanceOpts = {
  stroke?: {
    color: string,
    alpha?: number,
    width?: number,
    cap?: 'butt' | 'round' | 'square'
    dashOffset?: number
    join?: 'bevel' | 'round' | 'miter'
  }
}

export class LineAppearance {
  @observable stroke: {
    color: string,
    alpha: number,
    width: number,
    cap: 'butt' | 'round' | 'square'
    dashOffset: number
    join: 'bevel' | 'round' | 'miter'
  } = {
    color: '#000000',
    alpha: 1,
    width: 1,
    cap: 'butt',
    dashOffset: 0,
    join: 'miter',
  }

  constructor(opts: LineAppearanceOpts) {
    Object.assign(this.stroke, opts.stroke)
  }

  static of(opts: LineAppearanceOpts = {}): LineAppearance {
    return new LineAppearance(opts)
  }
}
