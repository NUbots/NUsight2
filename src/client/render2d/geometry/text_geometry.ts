import { observable } from 'mobx'

export class TextGeometry {
  @observable alignToView: boolean
  @observable fontFamily: string
  @observable maxWidth: number
  @observable text: string
  @observable textAlign: 'start' | 'middle' | 'end' | 'left' | 'right' | 'center'
  @observable textBaseline: 'top' | 'hanging' | 'middle' | 'alphabetic' | 'ideographic' | 'bottom'
  @observable x: number
  @observable y: number

  constructor(opts: TextGeometry) {
    this.alignToView = opts.alignToView
    this.fontFamily = opts.fontFamily
    this.maxWidth = opts.maxWidth
    this.text = opts.text
    this.textAlign = opts.textAlign
    this.textBaseline = opts.textBaseline
    this.x = opts.x
    this.y = opts.y
  }

  static of({
    alignToView = true,
    fontFamily = 'sans-serif',
    maxWidth = 0.5,
    text = '',
    textAlign = 'start',
    textBaseline = 'alphabetic',
    x = 0,
    y = 0,
  }: Partial<TextGeometry> = {}): TextGeometry {
    return new TextGeometry({
      alignToView,
      fontFamily,
      maxWidth,
      text,
      textAlign,
      textBaseline,
      x,
      y,
    })
  }
}
