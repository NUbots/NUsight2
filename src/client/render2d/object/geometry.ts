import { ArcGeometry } from '../geometry/arc_geometry'
import { ArrowGeometry } from '../geometry/arrow_geometry'
import { CircleGeometry } from '../geometry/circle_geometry'
import { LineGeometry } from '../geometry/line_geometry'
import { PathGeometry } from '../geometry/path_geometry'
import { PolygonGeometry } from '../geometry/polygon_geometry'
import { TextGeometry } from '../geometry/text_geometry'

export type Geometry =
  ArcGeometry
  | ArrowGeometry
  | CircleGeometry
  | LineGeometry
  | PathGeometry
  | PolygonGeometry
  | TextGeometry
