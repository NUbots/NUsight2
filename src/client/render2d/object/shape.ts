import { observable } from 'mobx'

import { Appearance } from '../appearance/appearance'
import { BasicAppearance } from '../appearance/basic_appearance'

import { Geometry } from './geometry'

export type ShapeOpts<T extends Geometry> = {
  appearance: Appearance
  geometry: T
}

export class Shape<T extends Geometry> {
  @observable appearance: Appearance
  @observable geometry: T

  constructor(opts: ShapeOpts<T>) {
    this.appearance = opts.appearance
    this.geometry = opts.geometry
  }

  static of<T extends Geometry>(geometry: T, appearance: Appearance = BasicAppearance.of()) {
    return new Shape<T>({
      appearance,
      geometry,
    })
  }
}
