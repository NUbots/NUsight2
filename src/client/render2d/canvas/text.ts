import { Transform } from '../../math/transform'
import { Vector2 } from '../../math/vector2'
import { TextGeometry } from '../geometry/text_geometry'
import { Shape } from '../object/shape'

import { applyAppearance } from './rendering'

export function renderText(
  ctx: CanvasRenderingContext2D,
  shape: Shape<TextGeometry>,
  world: Transform,
): void {
  const {
    x,
    y,
    text,
    fontSize,
    fontFamily,
    textAlign,
    textBaseline,
    worldAlignment,
    worldScale,
  } = shape.geometry

  ctx.font = `${fontSize} ${fontFamily}`
  ctx.textAlign = textAlign === 'middle' ? 'center' : textAlign
  ctx.textBaseline = textBaseline

  const position = Vector2.of(x, y)

  if (worldScale) {
    // Ensure the text is always rendered without rotation such that it is aligned with the screen.
    ctx.scale(1 / world.scale.x, 1 / world.scale.y)
  }

  if (worldAlignment) {
    ctx.rotate(-world.rotate)
  }

  ctx.translate(position.x, position.y)

  applyAppearance(ctx, shape.appearance)

  ctx.fillText(text, 0, 0)
}
