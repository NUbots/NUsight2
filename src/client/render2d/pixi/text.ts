import { createTransformer } from 'mobx-utils'
import { Text } from 'pixi.js'
import { TextStyle } from 'pixi.js'

import { Transform } from '../../math/transform'
import { Vector2 } from '../../math/vector2'
import { BasicAppearance } from '../appearance/basic_appearance'
import { TextGeometry } from '../geometry/text_geometry'
import { Shape } from '../object/shape'

import { applyAppearance } from './rendering'

export const renderText = createTransformer((shape: Shape<TextGeometry>): Text => {

  const { x, y, text, fontSize, fontFamily, textAlign, textBaseline, worldAlignment, worldScale } = shape.geometry

  if (shape.appearance instanceof BasicAppearance) {
    const t = new Text(text, {
      fontFamily,
      fontSize: fontSize,
      padding: 10,
      textBaseline: 'middle',
      fill: shape.appearance.fillStyle,
      align: textAlign,
    })
    t.x = x
    t.y = y
    t.anchor.x = 0.5
    t.anchor.y = 0.75 // I would have thought this should be 0.5, but 0.75 works to center text.
                      // As I have found no documentation to be sure of this, this might be wrong.
                      // but since it works it's what's here. If something goes wrong with text placement, look here.

    return t
  } else {
    throw Error('Pixi text only supports basic appearance')
  }
})
