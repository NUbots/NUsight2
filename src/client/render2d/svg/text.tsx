import { observer } from 'mobx-react'
import * as React from 'react'

import { Transform } from '../../math/transform'
import { TextGeometry } from '../geometry/text_geometry'
import { Shape } from '../object/shape'

import { toSvgProps, toSvgTransform } from './rendering'

type Props = { model: Shape<TextGeometry>, world: Transform }
export const Text = observer(({ model: { geometry, appearance }, world }: Props) => {
  const { x, y, fontFamily, fontSize, text, textAlign, textBaseline, worldAlignment, worldScale } = geometry

  const t = Transform.of({
    translate: { x, y },
    scale:  worldScale ? { x: 1 / world.scale.x, y: 1 / world.scale.y } : { x: 1, y: 1 },
    rotate: worldAlignment ? -world.rotate : 0,
  })

  // TODO (Houliston): Calculate fontSize such that the text fits within the given maxWidth
  // currently the font set here is correct for the dashboard view, but the canvas view is able to dynamically scale it
  return (
    <g transform={toSvgTransform(t)}>
      <text
        fontFamily={fontFamily}
        fontSize={fontSize}
        textAnchor={textAlign}
        dominantBaseline={textBaseline}
        {...toSvgProps(appearance)}
      >
        {text}
      </text>
    </g>
  )
})
