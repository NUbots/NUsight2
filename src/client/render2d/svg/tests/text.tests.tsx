import * as React from 'react'

import { Transform } from '../../../math/transform'
import { BasicAppearance } from '../../appearance/basic_appearance'
import { TextGeometry } from '../../geometry/text_geometry'
import { Shape } from '../../object/shape'
import { Text } from '../text'

describe('TextSVGRenderer', () => {
  it('renders', () => {
    const model = Shape.of(TextGeometry.of(), BasicAppearance.of())
    const world = Transform.of({
      scale: { x: -1, y: -1 },
      rotate: Math.PI / 4.0,
    })
    expect(<Text model={model} world={world}/>).toMatchSnapshot()
  })
})
