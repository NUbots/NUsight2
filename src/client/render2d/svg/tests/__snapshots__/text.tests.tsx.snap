// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextSVGRenderer renders 1`] = `
<_class
  model={
    Shape {
      "appearance": BasicAppearance {
        "fill": undefined,
        "stroke": undefined,
      },
      "geometry": TextGeometry {
        "fontFamily": "sans-serif",
        "fontSize": "1em",
        "text": "",
        "textAlign": "start",
        "textBaseline": "alphabetic",
        "worldAlignment": false,
        "worldScale": false,
        "x": 0,
        "y": 0,
      },
    }
  }
  world={
    Transform {
      "anticlockwise": true,
      "rotate": 0.7853981633974483,
      "scale": Object {
        "x": -1,
        "y": -1,
      },
      "translate": Object {
        "x": 0,
        "y": 0,
      },
    }
  }
/>
`;
